api/message.go

```diff
@@ -414,3 +417,58 @@ func toExternalMessages(msg []*model.Message) []*model.MessageExternal {
 	}
 	return res
 }
+
+// CreateMessageForClient creates a message, authentication via application-token is required.
+// swagger:operation POST /message_client message CreateMessageForClient
+//
+// Create a message.
+//
+// __NOTE__: This API ONLY accepts an application token as authentication.
+// ---
+// consumes: [application/json]
+// produces: [application/json]
+// security: [appTokenHeader: [], appTokenQuery: []]
+// parameters:
+// - name: body
+//   in: body
+//   description: the message to add
+//   required: true
+//   schema:
+//     $ref: "#/definitions/Message"
+// responses:
+//   200:
+//     description: Ok
+//     schema:
+//       $ref: "#/definitions/Message"
+//   400:
+//     description: Bad Request
+//     schema:
+//         $ref: "#/definitions/Error"
+//   401:
+//     description: Unauthorized
+//     schema:
+//         $ref: "#/definitions/Error"
+//   403:
+//     description: Forbidden
+//     schema:
+//         $ref: "#/definitions/Error"
+func (a *MessageAPI) CreateMessageForClient(ctx *gin.Context) {
+	message := model.MessageExternal{}
+	if err := ctx.Bind(&message); err == nil {
+		token := auth.GetTokenID(ctx)
+		client, err := a.DB.GetClientByToken(token)
+		if success := successOrAbort(ctx, 500, err); !success {
+			return
+		}
+
+		message.ApplicationID = 0
```

Comment:
Hey, Thanks for your contribution.

Adding a send message api with client authentication is not possible with the current design. Each message must have an application assigned otherwise the ui don't know which icon to show for the message.

Like written in https://github.com/gotify/server/issues/186 the /message api should be extended to accept a client_id query parameter to only send the message to a specific client. The authentication should still be done with an application token.

How the client_id parameter should behave isn't clearly defined in https://github.com/gotify/server/issues/186 this still needs some discussion before implementing this feature. See my last comment.